[raysgd] Custom training operator (#7211)
ray-project/ray: doc/source/raysgd/raysgd_pytorch.rst

        num_replicas=100,
        use_gpu=True)

See the documentation on the PyTorchTrainer here: :ref:`ref-pytorch-trainer`.

.. _raysgd-custom-training:

Custom Training and Validation (Operators)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> ``PyTorchTrainer`` allows you to run a custom training and validation loops in parallel on each worker, providing a flexibility similar to using PyTorch natively.

Author: ```suggestion
``PyTorchTrainer`` allows you to run a custom training and validation loops in parallel on each worker, providing a flexible interface similar to using PyTorch natively.
```

This is done via the :ref:`ref-pytorch-operator` interface.

> For both training and validation, there are two granularities that you can provide customization - per epoch or per batch. These correspond to ``train_batch``,

Reviewer: @maximsmol FYI you can click on the line number to suggest changes inline

> For both training and validation, there are two granularities that you can provide customization - per epoch or per batch. These correspond to ``train_batch``,

Author: ```suggestion
For both training and validation, there are two granularities that you can provide customization - per epoch and per batch. These correspond to ``train_batch``,
```

> ``train_epoch``, ``validate``, and ``validate_batch``. Other useful methods to override include ``setup``, ``save`` and ``restore``. You can use this class

Author: ```suggestion
``train_epoch``, ``validate``, and ``validate_batch``. Other useful methods to override include ``setup``, ``save`` and ``restore``. You can use these
```

to manage state (like a classifier neural network for calculating inception score, or a heavy tokenizer).

Providing a custom operator is necessary if creator functions return multiple models, optimizers, or schedulers.

> Below is a partial example of overriding ``train_batch``.

Author: ```suggestion
Below is a partial example of a custom ``TrainingOperator`` that provides a ``train_batch`` implementation for a Deep Convolutional GAN.
```

.. code-block:: python

    from ray.util.sgd.pytorch import TrainingOperator

    class GANOperator(TrainingOperator):
        def setup(self, config):
            """Custom setup for this operator.

            Args:
                config (dict): Custom configuration value to be passed to
                    all creator and operator constructors. Same as ``self.config``.
            """
            self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

        def train_batch(self, batch, batch_info):
            """Trains on one batch of data from the data creator.

            Args:
                batch: One item of the validation iterator.
                batch_info (dict): Information dict passed in from ``train_epoch``.

            Returns:
                A dict of metrics. Defaults to "loss" and "num_samples",
                    corresponding to the total number of datapoints in the batch.
            """
            real_label = 1
            fake_label = 0
            discriminator, generator = self.models
            optimD, optimG = self.optimizers

>             discriminator.zero_grad()

Reviewer: comments with high-level descriptions of what's happening in each of these blocks would be helpful

Author: fixed.

            optimD.step()

            generator.zero_grad()
            label.fill_(real_label)
            output = discriminator(fake).view(-1)
            errG = self.criterion(output, label)
            errG.backward()
            optimG.step()

            is_score, is_std = self.inception_score(fake)

            return {
                "loss_g": errG.item(),
                "loss_d": errD.item(),
                "inception": is_score,
                "num_samples": batch_size
            }

    trainer = PyTorchTrainer(
            model_creator,
            data_creator,
            optimizer_creator,
            nn.BCELoss,
            training_operator_cls=GANOperator,
            num_replicas=num_replicas,
            config=config,
            use_gpu=True,
            batch_size=128)

    for i in range(5):
        stats = trainer.train()
        print(stats)

See the `DCGAN example <https://github.com/ray-project/ray/blob/master/python/ray/util/sgd/pytorch/examples/dcgan.py>`__ for an end to end example. It constructs two models and two optimizers and uses a custom training operator to provide a non-standard training loop.

Training/Validation
-------------------

> Now that the trainer is constructed, you'll naturally want to train the model.

Reviewer: ```suggestion
Now that the trainer is constructed, here's how to train the model.
```
"naturally" sounds like fluff, always good to use simple language where possible for things like these IMO

> Now that the trainer is constructed, you'll naturally want to train the model.

Author: done

.. code-block:: python

    for i in range(10):
        metrics = trainer.train()
        val_metrics = trainer.validate()

> Each ``train`` takes one pass over the training data, and each ``validate`` call runs the model on the validation data passed in by the ``data_creator``. Provide a custom training operator (:ref:`raysgd-custom-training`) to calculate custom metrics or customize the training/validation process.

Reviewer: ```suggestion
Each ``train`` call makes one pass over the training data and each ``validate`` call runs the model on the validation data passed in by the ``data_creator``. Provide a custom training operator (:ref:`raysgd-custom-training`) to calculate custom metrics or customize the training/validation process.
```

Initialization Functions
------------------------

> Use the ``initialization_hook`` parameter to initialize state on each worker process when they are started. This is useful when setting an environment variable or downloading data:

Author: done.

        }
    )

Note that if using a custom training operator (:ref:`raysgd-custom-training`), you will need to manage loss scaling manually.

Distributed Multi-node Training
-------------------------------

> You can scale out your training onto multiple nodes without making any modifications to your training code. To train across a cluster, simply make sure that the Ray cluster is started. You can start a Ray cluster `via the Ray cluster launcher <autoscaling.html>`_ or `manually <using-ray-on-a-cluster.html>`_.

Reviewer: ```suggestion
You can scale your training to multiple nodes without making any modifications to the training code. To train across a cluster, simply make sure that the Ray cluster is started. You can start a Ray cluster `via the Ray cluster launcher <autoscaling.html>`_ or `manually <using-ray-on-a-cluster.html>`_.
```

Reviewer: This was a bit confusing to me because you say "simply make sure that the Ray cluster is started" then end the sentence without indicating that you have to change the code to connect to it. I thought that was going to magically work - readers might stop there and think that too

Author: fixed.

In your program, you'll need to connect to this cluster via ``ray.init``:

.. code-block:: python

>     ray.init(address="auto")  # or a specific redis address.

Reviewer: ... "of the form <ip-address:port>"

Author: fixed

Simultaneous Multi-model Training
---------------------------------

> In certain scenarios such as training GANs, you may want to use multiple models in the training loop. You can do this in the ``PyTorchTrainer`` by allowing the ``model_creator``, ``optimizer_creator``, and ``scheduler_creator`` to return multiple values. Provide a custom TrainingOperator (:ref:`raysgd-custom-training`) to train across multiple models.

Reviewer: ```suggestion
In certain scenarios, such as training GANs, you may want to use multiple models in the training loop. You can do this in the ``PyTorchTrainer`` by allowing the ``model_creator``, ``optimizer_creator``, and ``scheduler_creator`` to return multiple values. Provide a custom TrainingOperator (:ref:`raysgd-custom-training`) to train across multiple models.
```
